app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt

```diff
@@ -103,141 +127,171 @@ fun ReaderScreen(info: BaseGalleryInfo, page: Int = -1, navigator: DestinationsN
+                    modifier = Modifier.align(Alignment.BottomCenter).navigationBarsPadding(),
+                )
+            }
+        }
+        ReaderAppBars(
+            visible = appbarVisible,
+            isRtl = readingMode == ReadingModeType.RIGHT_TO_LEFT,
+            showSeekBar = showSeekbar,
+            currentPage = syncState.sliderValue,
+            totalPages = pageLoader.size,
+            onSliderValueChange = { syncState.sliderScrollTo(it + 1) },
+            onClickSettings = {
+                launch {
+                    dialog { cont ->
+                        fun dispose() = cont.resume(Unit)
+                        var isColorFilter by remember { mutableStateOf(false) }
+                        val scrim by animateColorAsState(
+                            targetValue = if (isColorFilter) Color.Transparent else BottomSheetDefaults.ScrimColor,
+                            label = "ScrimColor",
+                        )
+                        ModalBottomSheet(
+                            onDismissRequest = { dispose() },
+                            modifier = Modifier.windowInsetsPadding(WindowInsets.safeDrawing.only(WindowInsetsSides.Top)),
+                            // Yeah, I know color state should not be read here, but we have to do it...
+                            scrimColor = scrim,
+                            dragHandle = null,
+                            contentWindowInsets = { WindowInsets(0) },
+                        ) {
+                            SettingsPager(modifier = Modifier.fillMaxSize()) { page ->
+                                isColorFilter = page == 2
+                                appbarVisible = !isColorFilter
                             }
                         }
                     }
-                },
-            )
+                }
+            },
+        )
+    }
+}
+
+context(Context, DialogState, DestinationsNavigator)
+private suspend fun preparePageLoader(args: ReaderScreenArgs) = when (args) {
+    is ReaderScreenArgs.Gallery -> {
+        val info = args.info
+        val page = args.page
+        val archive = DownloadManager.getDownloadInfo(info.gid)?.archiveFile
+        if (archive != null) {
+            ArchivePageLoader(archive, info.gid, page)
+        } else {
+            EhPageLoader(info, page)
+        }
+    }
+    is ReaderScreenArgs.Archive -> {
+        ArchivePageLoader(args.uri.toOkioPath()) { invalidator ->
+            kotlin.runCatching {
+                awaitInputText(
+                    title = getString(R.string.archive_need_passwd),
+                    hint = getString(R.string.archive_passwd),
+                ) { text ->
+                    ensure(text.isNotBlank()) { getString(R.string.passwd_cannot_be_empty) }
+                    ensure(invalidator(text)) { getString(R.string.passwd_wrong) }
+                }
+            }.onFailure {
+                withUIContext { popBackStack() }
```

Comment:
Let ArchivePageloader not ready if no correct password provided, and merge the error path to outside one.

Reply:
Well I was considering if we want to consider cancellation an error (and show snackbar).

Reply:
> Well I was considering if we want to consider cancellation an error (and show snackbar).

cancellation should not be error, and awaitReady should propagate it then we show correct error msg

Reply:
We can't let cancellation propagate as it will cancel the block so the content of `Deferred` won't be executed.

Reply:
> We can't let cancellation propagate as it will cancel the block so the content of `Deferred` won't be executed.

Let's try `awaitReady` and deliver error content.